core/src/main/java/com/linecorp/armeria/server/WebSocketSessionHandler.java

```diff
@@ -0,0 +1,126 @@
+        super.channelUnregistered(ctx);
+        req.close(ClosedSessionException.get());
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (evt instanceof InitiateConnectionShutdown) {
+            encoder.keepAliveHandler().disconnectWhenFinished();
+            return;
+        }
+        ctx.fireUserEventTriggered(evt);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        if (msg instanceof DecodedHttpRequest) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+        if (!(msg instanceof ByteBuf)) {
+            return;
+        }
+        try {
+            final ByteBuf data = (ByteBuf) msg;
+            final int dataLength = data.readableBytes();
+            if (dataLength != 0) {
+                req.increaseTransferredBytes(dataLength);
+                final long maxContentLength = req.maxRequestLength();
+                final long transferredLength = req.transferredBytes();
+                if (maxContentLength > 0 && transferredLength > maxContentLength) {
+                    final ContentTooLargeException cause =
+                            ContentTooLargeException.builder()
+                                                    .maxContentLength(maxContentLength)
+                                                    .contentLength(req.headers())
+                                                    .transferred(transferredLength)
+                                                    .build();
+                    if (encoder.isResponseHeadersSent(req.id(), 1)) {
+                        encoder.writeReset(req.id(), 1, Http2Error.PROTOCOL_ERROR);
+                    } else {
+                        encoder.writeErrorResponse(req.id(), 1, serviceConfig, req.headers(),
+                                                   HttpStatus.REQUEST_ENTITY_TOO_LARGE, null, null);
+                    }
+                    req.abortResponse(
+                            HttpStatusException.of(HttpStatus.REQUEST_ENTITY_TOO_LARGE, cause),
+                            true);
+                    return;
+                }
+
+                if (req.isOpen()) {
+                    req.write(HttpData.wrap(data.retain()));
+                }
+            }
+        } finally {
+            ReferenceCountUtil.release(msg);
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (msg instanceof HttpResponse) {
+            final HttpResponse response = (HttpResponse) msg;
+            final HttpResponseStatus status = response.status();
+            ctx.write(msg, promise);
+            if (status == HttpResponseStatus.SWITCHING_PROTOCOLS) {
+                ctx.pipeline().remove(HttpServerCodec.class);
+            }
+            return;
+        }
+
+        if (msg instanceof HttpContent) {
```

Comment:
Is it feasible to add `WebSocketHttp1ObjectEncoder`? If so, we may directly deliver `ByteBuf` to the Netty channel buffer without unwrapping `HttpContent`.